### PR TITLE
vmm: Ensure PIO/MMIO exits complete before pausing only for KVM

### PR DESCRIPTION
**Pull request by russell-islam**
Looks like this code guarded by kvm feature flag does not work properly. When we build with both kvm and mshv restore hangs indefinitely.

### CODE REVIEW

**Comment by russell-islam** [comment]
Fixes https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5037

File: vmm/src/cpu.rs [hunk 1]
```diff
@@ -811,6 +811,8 @@ impl CpuManager {
     ) -> Result<()> {
         let reset_evt = self.reset_evt.try_clone().unwrap();
         let exit_evt = self.exit_evt.try_clone().unwrap();
+        #[cfg(feature = "kvm")]
+        let on_kvm: bool = HypervisorType::Kvm == self.hypervisor_type;
```
Comment (likebreath) [inline vmm/src/cpu.rs:815]: Nit: it would be cleaner if you do:
```suggestion
        #[cfg(feature = "kvm")]
        let hypervisor_type = self.hypervisor_type;
```

File: vmm/src/cpu.rs [hunk 2]
```diff
@@ -919,12 +921,14 @@ impl CpuManager {
...
+                                    if on_kvm {
+                                        vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(true);
+                                        if !matches!(vcpu.lock().unwrap().run(), Ok(VmExit::Ignore)) {
+                                            error!("Unexpected VM exit on \"immediate_exit\" run");
+                                            break;
+                                        }
+                                        vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(false);
                                     }
-                                    vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(false);
                                 }
```
Comment (likebreath) [inline vmm/src/cpu.rs:932]: Nit: it would be cleaner if you do:

```
#[cfg(feature = "kvm")]
if hypervisor_type == HypervisorType::Kvm {
    vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(true);
    if !matches!(vcpu.lock().unwrap().run(), Ok(VmExit::Ignore)) {
        error!("Unexpected VM exit on \"immediate_exit\" run");
        break;
    }
    vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(false);
}
```

**Review by likebreath — CHANGES_REQUESTED** [review]
The PR changes look good. Only two nits below.  But I think the commit message can be improved as the PR is not about "set_immediate_exit". 

You can try something like: `vmm: Ensure PIO/MMIO exits complete before pausing only for KVM`. Also, I would mention in the commit message body that "MSHV does not require to ensure MMIO/PIO exits complete before pausing" (if this statement is true and is the reason why you need to make the changes).

**Comment by likebreath** [comment]
> Fixes #5037

Please also include this to the commit message body

**Comment by russell-islam** [comment]
Thank you @likebreath. Applied your suggestions. :)

File: vmm/src/cpu.rs [hunk 3]
```diff
@@ -919,12 +921,14 @@ impl CpuManager {
 
                                 #[cfg(feature = "kvm")]
                                 {
-                                    vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(true);
-                                    if !matches!(vcpu.lock().unwrap().run(), Ok(VmExit::Ignore)) {
-                                        error!("Unexpected VM exit on \"immediate_exit\" run");
-                                        break;
+                                    if on_kvm {
```
Comment (sboeuf) [inline vmm/src/cpu.rs:924]: Out of curiosity, why is this needed given it's already behind a KVM feature gate?

File: vmm/src/cpu.rs [hunk 4]
```diff
@@ -811,6 +811,8 @@ impl CpuManager {
     ) -> Result<()> {
         let reset_evt = self.reset_evt.try_clone().unwrap();
         let exit_evt = self.exit_evt.try_clone().unwrap();
+        #[cfg(feature = "kvm")]
```
Comment (sboeuf) [inline vmm/src/cpu.rs:814]: Please fix both your commit title and your commit message, replacing `exists` with `exits`.

File: vmm/src/cpu.rs (hunk 3 above)
Comment (rbradford) [reply vmm/src/cpu.rs:924]: This issue was found when running on a dual binary (i.e. both kvm and mshv compiled in) running on mshv. `set_immediate_exit` is only implemented on kvm (and wouldn't be needed on mshv due to the different method of decoding mmio exits.)

Comment (sboeuf) [reply vmm/src/cpu.rs:924]: Gotcha!

**Comment by liuw** [comment]
@NunoDasNeves perhaps this can be implemented by a cancel API?

(This is not to say we should do it immediately, but just a good use case for that)

**Comment by rbradford** [comment]
> @NunoDasNeves perhaps this can be implemented by a cancel API?
> 
> (This is not to say we should do it immediately, but just a good use case for that)

From my understanding of the MSHV MMIO handling the behaviour we get get with KVM_IMMEDIATE_EXIT isn't necessary. It's only necessary on KVM because MMIO handling is only actually completed on KVM once the KVM_RUN call is re-entered. Hence why we must drop into KVM_RUN when pausing to complete a pending MMIO operation if we find we need to pause.

**Comment by liuw** [comment]
> > @NunoDasNeves perhaps this can be implemented by a cancel API?
> > (This is not to say we should do it immediately, but just a good use case for that)
> 
> From my understanding of the MSHV MMIO handling the behaviour we get get with KVM_IMMEDIATE_EXIT isn't necessary. It's only necessary on KVM because MMIO handling is only actually completed on KVM once the KVM_RUN call is re-entered. Hence why we must drop into KVM_RUN when pausing to complete a pending MMIO operation if we find we need to pause.

Yes that's my understanding too.

File: vmm/src/cpu.rs (hunk 2 above)
Comment (likebreath) [reply vmm/src/cpu.rs:932]: @russell-islam Please don't forget to address two nits here. Thanks.

File: vmm/src/cpu.rs (hunk 4 above)
Comment (rbradford) [reply vmm/src/cpu.rs:814]: @russell-islam You still missed a typo in the commit message body.
> MSHV does not require to ensure MMIO/PIO exists complete
before pausing. This patch makes sure the above requirement
by checking the hypervisor type run-time.

File: vmm/src/cpu.rs (hunk 3 above)
Comment (rbradford) [inline vmm/src/cpu.rs:924]: You can drop the first commit in the series and simplify this all by just doing `if matches!(self.hypervisor_type, HypervisorType::Kvm)`

File: vmm/src/cpu.rs [hunk 5]
```diff
@@ -919,12 +921,14 @@ impl CpuManager {
 
                                 #[cfg(feature = "kvm")]
                                 {
-                                    vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(true);
-                                    if !matches!(vcpu.lock().unwrap().run(), Ok(VmExit::Ignore)) {
-                                        error!("Unexpected VM exit on \"immediate_exit\" run");
-                                        break;
+                                    if matches!(hypervisor_type, HypervisorType::Kvm) {
```
Comment (likebreath) [inline vmm/src/cpu.rs:924]: @russell-islam My main point was you can move `#[cfg(feature = "kvm")]` right before the `if matches!(hypervisor_type, HypervisorType::Kvm)` statement, which would avoid the extra parentheses and make the patch smaller/cleaner (4 lines of changed in total).

Comment (rbradford) [reply vmm/src/cpu.rs:924]: @likebreath Sorry, I missed your comment since there was an already overlapping comment.  You are right a layer of indentation/grouping could be removed. Could you create a follow-on PR?

Comment (russell-islam) [reply vmm/src/cpu.rs:924]: I will create a follow-up. Sorry for the delay!!
